Batch import metadata

mathiasflorin opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Dependency Update
- [ ] Release

## Description
Use batch import instead of loading one large file. 


<img width="561" alt="image" src="https://github.com/user-attachments/assets/a30b9cda-fe26-4fe3-968e-dcb5e27e5255">


## Related Tickets & Documents


- Related Issue #
- Closes #

## Linting

Have you done linting by issuing `./build.sh lint` command?
- [X] Yes
- [ ] No, I need help with linting


## Testing

Have you run testing by issuing `./build.sh unit` command?
- [ ] Yes
- [X] No, but I tested the export and import

## Documentation
Have you updated the [README](../README.md) appropriate for this PR?
- [ ] Yes
- [X] No, README does not need any changes for this PR

## Changelog

Have you updated the [CHANGELOG](../CHANGELOG.md) with the changes you are making in this PR?
Please use Added, Removed, and/or Changed subsections under the [Unreleased](../CHANGELOG.md#unreleased) section.
- [X] Yes
- [ ] No, I need help

## Terms of Contribution

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

mathiasflorin commented on assets/dags/mwaa_dr/v_2_4/dr_factory.py:
```diff
@@ -182,6 +182,7 @@ def dag_run(self, model: DependencyModel[BaseTable]) -> BaseTable:
                 "start_date",
                 "state",
             ],
+            export_filter="dag_id != 'backup_metadata'",
```
Filters backup_metadata to ensure the scheduler does not resume backup_metadata with the status running during batch imports of metadata. If the scheduler resumes, it will overwrite the metadata while the import is running.

crupakheti commented:
Appreciate your code contribution, @mathiasflorin ! I plan to work on this PR sometimes next week. Thank you!

crupakheti commented:
Meanwhile, please make sure all tests are passing and dont worry about Publish to TestPyPi failure. Thank you!

mathiasflorin commented:
> Meanwhile, please make sure all tests are passing and dont worry about Publish to TestPyPi failure. Thank you!

I have modified the tests to accommodate the batch import approach. Additionally, I have added a new test to verify that the commit is called twice when the data to be loaded exceeds the batch size.

mathiasflorin commented:
Hi @crupakheti, I've already added an additional test for the batch import. However, I can add more tests if you'd like to maintain the code coverage at 100%.

crupakheti commented:
Thank you, @mathiasflorin ! Yes, please. Let's try to maintain the coverage at 100%.

mayushko26 reviewed (commented):
Great change! These backup files could get very large, so batching the file-to-database import is a great improvement. I have two overall comments:

1 - I'm wondering if you were able to somewhat load-test the new import logic with a particularly large backup file, as this change is targeted at improving the scalability of the import step. For example, one thing that comes to mind is that frequent COPY-commit commands in quick succession could be expensive or high-latency due to row locking and transaction overhead. Some of these can be mitigated by chunking the commits themselves. And maybe a question to @crupakheti, is there benchmarking for some of the more resource-heavy operations in the MWAA DR tool?

2 - I want to call out that this doesn't close #31; as that issue relates to the export (backup) dag, while this change is focused on chunking the import (restore) dag.

mathiasflorin commented:
> Great change! These backup files could get very large, so batching the file-to-database import is a great improvement. I have two overall comments:
> 
> 1 - I'm wondering if you were able to somewhat load-test the new import logic with a particularly large backup file, as this change is targeted at improving the scalability of the import step. For example, one thing that comes to mind is that frequent COPY-commit commands in quick succession could be expensive or high-latency due to row locking and transaction overhead. Some of these can be mitigated by chunking the commits themselves. And maybe a question to @crupakheti, is there benchmarking for some of the more resource-heavy operations in the MWAA DR tool?

I have tested this code with CSV files containing several million records. As I will be going on holiday soon, would it make sense to merge this pull request now and address your suggested improvements in a follow-up feature pull request?
> 2 - I want to call out that this doesn't close #31; as that issue relates to the export (backup) dag, while this change is focused on chunking the import (restore) dag.
Sorry for the confusion. You are correct that this change does not close #31. I will update the pull request description accordingly. This change addresses the memory pressure issue during import rather than export.

mathiasflorin commented on assets/dags/mwaa_dr/framework/model/base_table.py:
```diff
@@ -347,19 +347,34 @@ def restore(self, **context):
+            insert_counter = 0
+            with backup_file as file:
+                batch = []
+                for line in file:
```
I believe multiple commits are necessary to commit smaller batches of data instead of one large operation. Please feel free to add your code suggestions, but I will not be able to test them as I will be going on holiday soon. I am happy to give ownership of this pull request to you if you'd like. I am working on getting the code coverage back to 100%, but I plan to move on to another topic afterward.

mathiasflorin replied:
<img width="577" alt="image" src="https://github.com/user-attachments/assets/6af3b924-51e0-4c50-aa12-61b919eb82c3">

Are you happy to add your proposal and test the code?

mathiasflorin replied:
Yes, I have implemented your suggestion to use `itertools.islice`. Thank you for the recommendation.

mayushko26 commented:
I will likely not have time to get to it this week, but I am +1 for biasing to merge this, and creating a separate Issue to track this improvement proposal. @crupakheti What do you think?

crupakheti commented:
Thank you for your contribution! Your code has been merged into `main` through PR #38. 🙏
